### PR TITLE
Iterator with timestamp

### PR DESCRIPTION
**Pull request by riversand963**
Preliminary support for iterator with user timestamp. Current implementation does not consider merge operator and reverse iterator. Auto compaction is also disabled in unit tests.

Create an iterator with timestamp.
```
...
read_opts.timestamp = &ts;
auto* iter = db->NewIterator(read_opts);
// target is key without timestamp.
for (iter->Seek(target); iter->Valid(); iter->Next()) {}
for (iter->SeekToFirst(); iter->Valid(); iter->Next()) {}
delete iter;
read_opts.timestamp = &ts1;
// lower_bound and upper_bound are without timestamp.
read_opts.iterate_lower_bound = &lower_bound;
read_opts.iterate_upper_bound = &upper_bound;
auto* iter1 = db->NewIterator(read_opts);
// Do Seek or SeekToFirst()
delete iter1;
```

Test plan (dev server)
```
$make check
```

Simple benchmarking (dev server)
1. The overhead introduced by this PR even when timestamp is disabled.
key size: 16 bytes
value size: 100 bytes
Entries: 1000000
Data reside in main memory, and try to stress iterator.
Repeated three times on master and this PR.
- Seek without next
```
./db_bench -db=/dev/shm/rocksdbtest-1000 -benchmarks=fillseq,seekrandom -enable_pipelined_write=false -disable_wal=true -format_version=3
```
master: 159047.0 ops/sec
this PR: 158922.3 ops/sec (2% drop in throughput)
- Seek and next 10 times
```
./db_bench -db=/dev/shm/rocksdbtest-1000 -benchmarks=fillseq,seekrandom -enable_pipelined_write=false -disable_wal=true -format_version=3 -seek_nexts=10
```
master: 109539.3 ops/sec
this PR: 107519.7 ops/sec (2% drop in throughput)

### CODE REVIEW

**Comment by riversand963** [comment]
cc @siying @maysamyabandeh and @hliu18. Comments welcome, though it's still WIP.

**Comment by wolfkdy** [comment]
I have these confusions
1） if readOptions have snapshot(lsn) and timestamp both set, will Get/Seek return the records which satisfy both restrictions?
2） snapshot(lsn) may pin records from being compacted, will timestamp also have similar behavior?
that said, will there be some api to set an "oldest-timestamp" to the database, and only records older than oldest-timestamp are possible to be purged?

**Comment by riversand963** [comment]
Thanks @wolfkdy for your interest! I have tried to address your questions below.

> I have these confusions， hoping for your patient explanations @riversand963
> 1） Does timestamp option mean a strict match or a fuzzy match?
> that said, if I have a key with timestamp=10, will a Get/Seek with timestamp = 11 see it?
The key is visible.
> 2）whose duty is it to guarantee that，to the same key, the timestamp order is the same as the commit order ?
Application/transaction manager above `rocksdb::DB`. For example, `TransactionDB` or other application provides this guarantee.
We have thought about relaxing this restriction. However, it will not only complicate implementation, but also introduce lookup/range scan performance issues. Therefore, we stick to this restriction at the moment.
> 3） if readOptions have snapshot(lsn) and timestamp both set, will Get/Seek return the records which satisfy both restrictions?
Get/Iterator should see records that satisfy both. Note that timestamp assignment and sequence number assignment do not happen atomically. If an application tries to perform Get() or range scan with read_options.timestamp = ts and read_options.snapshot = seq, the application should make sure no write succeeds with ts' < ts is assigned seq' > seq. Or put it another way, if there is inflight write with ts' and seq1, the application should not try to read with ts > ts' but seq < seq'.
> 4） snapshot(lsn) may pin records from being compacted, will timestamp also have similar behavior?
> that said, will there be some api to set an "oldest-timestamp" to the database, and only records older than oldest-timestamp are possible to be purged?
Yes, there will an API like this.

**Comment by wolfkdy** [comment]
@riversand963  thanks for your detailed explanations. Elegant and reasonable design!
Base on your nice work, I have some ideas about the txn-manager layer.
I’d like to write a prototype first, and leave some apis blank until the underneath layer has fulfilled them.
What do you think about it?  Can we work together about this, because I think it's a long way to go.

Due to the timezone, we can discuss in  wechat

**Comment by BradBarnich** [comment]
This is great. I had been working adding timestamp support to iterators on my own, but I like this design better than what I came up with.

I replaced my implementation with this, but kept my tests. There were a few things I had to add back to keep the tests working. They might be for features you are intentionally not working on yet, like reverse iteration, but I thought it might be helpful to share: https://github.com/BradBarnich/rocksdb/commit/5531aa330a8345dfb8ccda56de4a0b68a3f695b8#diff-0465a3d0601c0cd6f05a6d0e9bfabd36

**Comment by riversand963** [comment]
> This is great. I had been working adding timestamp support to iterators on my own, but I like this design better than what I came up with.
> 
> I replaced my implementation with this, but kept my tests. There were a few things I had to add back to keep the tests working. They might be for features you are intentionally not working on yet, like reverse iteration, but I thought it might be helpful to share: [BradBarnich@5531aa3#diff-0465a3d0601c0cd6f05a6d0e9bfabd36](https://github.com/BradBarnich/rocksdb/commit/5531aa330a8345dfb8ccda56de4a0b68a3f695b8#diff-0465a3d0601c0cd6f05a6d0e9bfabd36)

Thanks @BradBarnich for sharing. Yes, this PR is far from complete timestamp for iterators. Reverse iteration, merge operator, compaction, etc. are not in the scope of this PR and will be addressed in the future.

File: include/rocksdb/iterator.h [hunk 1]
```diff
@@ -108,6 +108,11 @@ class Iterator : public Cleanable {
   //   Get the user-key portion of the internal key at which the iteration
   //   stopped.
   virtual Status GetProperty(std::string prop_name, std::string* prop);
+
+  virtual Slice utimestamp() const {
```
Comment (siying) [inline include/rocksdb/iterator.h]: I guess "u" stands for "user"? Is "timestamp" good enough?

File: table/block_based/block.cc [hunk 2]
```diff
@@ -438,8 +429,8 @@ void DataBlockIter::SeekForPrev(const Slice& target) {
     return;
   }
   SeekToRestartPoint(index);
-  // Linear search (within restart block) for first key >= seek_key
 
+  // Linear search (within restart block) for first key >= seek_key
```
Comment (siying) [inline table/block_based/block.cc]: Nothing really changed in this file? If it is not too hard, I suggest we separate code cleaning up from this unrelated feature.

File: table/block_based/block.h [hunk 3]
```diff
@@ -559,13 +557,13 @@ class IndexBlockIter final : public BlockIter<IndexValue> {
...
+  void Prev() override;
 
-  virtual void Next() override;
+  void Next() override;
 
-  virtual void SeekToFirst() override;
+  void SeekToFirst() override;
 
-  virtual void SeekToLast() override;
+  void SeekToLast() override;
```
Comment (siying) [inline table/block_based/block.h]: Changes from this file are code cleaning up and are not related to the PR too? If possible, please move it to another PR.

File: table/iterator_wrapper.h [hunk 4]
```diff
@@ -83,8 +90,16 @@ class IteratorWrapperBase {
...
+  void SeekToFirst() {
+    assert(iter_);
+    iter_->SeekToFirst();
+    Update();
+  }
+  void SeekToLast() {
+    assert(iter_);
+    iter_->SeekToLast();
+    Update();
+  }
```
Comment (siying) [inline table/iterator_wrapper.h]: Again. This file doesn't really change, right?

File: db/arena_wrapped_db_iter.cc [hunk 5]
```diff
@@ -18,6 +18,33 @@
...
+        new (saved_addr) Slice(buf, bound->size() + ts->size());
+    return bound_and_ts;
+  };
+  if (iter_lb) {
+    ret.iterate_lower_bound = copy_iter_bounds_and_ts(iter_lb);
+  }
+  if (iter_ub) {
+    ret.iterate_upper_bound = copy_iter_bounds_and_ts(iter_ub);
+  }
+  return ret;
```
Comment (siying) [inline db/arena_wrapped_db_iter.cc]: Can you explain why we need to make a copy to the upper and lower bounds?
I'm not sure it works with a scenario that upper and lower bound are changed in Seek().

File: db/db_iter.cc [hunk 6]
```diff
@@ -1060,10 +1077,20 @@ bool DBIter::TooManyInternalKeysSkipped(bool increment) {
   return false;
 }
 
-bool DBIter::IsVisible(SequenceNumber sequence) {
+bool DBIter::IsVisible(SequenceNumber sequence, const Slice& ts) {
+  int cmp_ts = timestamp_ub_ != nullptr
+                   ? user_comparator_.CompareTimestamp(ts, *timestamp_ub_)
+                   : 0;
+  if (cmp_ts < 0) {
+    return false;
```
Comment (siying) [inline db/db_iter.cc]: If ts in key < read_options.ts, it should be visible?

File: db/db_iter.cc [hunk 7]
```diff
@@ -1060,10 +1077,20 @@ bool DBIter::TooManyInternalKeysSkipped(bool increment) {
...
+  } else if (cmp_ts > 0) {
```
Comment (siying) [inline db/db_iter.cc]: If key > read_options.ts, it should be invisible?

File: db/db_iter.cc [hunk 8]
```diff
@@ -1060,10 +1077,20 @@ bool DBIter::TooManyInternalKeysSkipped(bool increment) {
...
+    assert(sequence <= sequence_);
```
Comment (siying) [inline db/db_iter.cc]: Can you explain why this assertion?

File: table/block_based/block.h (hunk 3 above)
Comment (riversand963) [reply table/block_based/block.h]: Thanks @siying for the suggestion, I have moved code cleaning to https://github.com/facebook/rocksdb/pull/6440.

File: include/rocksdb/iterator.h (hunk 1 above)
Comment (riversand963) [reply include/rocksdb/iterator.h]: Sadly no. class `TtlIterator` has a member function `uint32_t timestamp()` already.

File: db/db_iter.cc (hunk 6 above)
Comment (riversand963) [reply db/db_iter.cc]: Here, cmp_ts < 0 means key.ts is *after* read_opts.timestamp. The timestamp comparator orders newer key to be "smaller".

File: db/db_iter.cc (hunk 8 above)
Comment (riversand963) [reply db/db_iter.cc]: `cmp_ts>0` -> `key.ts > ts_ub` -> key.ts is before ts_ub. We currently assume timestamp order is the same as sequence number order. Thus, the key.seq is smaller than (before) snapshot.seq.

File: db/arena_wrapped_db_iter.cc (hunk 5 above)
Comment (riversand963) [reply db/arena_wrapped_db_iter.cc]: This is a valid concern which I did not consider. Will revise the implementation.

File: table/block_based/block.cc (hunk 2 above)
Comment (riversand963) [reply table/block_based/block.cc]: Done.

File: table/block_based/block.h (hunk 3 above)
Comment (riversand963) [reply table/block_based/block.h]: Done.

File: table/iterator_wrapper.h (hunk 4 above)
Comment (riversand963) [reply table/iterator_wrapper.h]: Done.

File: db/arena_wrapped_db_iter.cc (hunk 5 above)
Comment (riversand963) [reply db/arena_wrapped_db_iter.cc]: @siying. Updated the PR and removed the copy of iteration bounds.

File: db/db_iter.cc (hunk 6 above)
Comment (riversand963) [reply db/db_iter.cc]: On second thought, I think the logic may be a little twisted here. Let me change it.

File: include/rocksdb/iterator.h (hunk 1 above)
Comment (siying) [reply include/rocksdb/iterator.h]: TtlIterator is an internal interface, which we can freely rename. Right?
This function may stay there forever, so it worth some time to come up with a good name.

File: util/user_comparator_wrapper.h [hunk 9]
```diff
@@ -58,6 +58,16 @@ class UserComparatorWrapper final : public Comparator {
...
   }
 
+  int CompareTimestamp(const Slice& ts1, const Slice& ts2) const override {
+    return user_comparator_->CompareTimestamp(ts1, ts2);
+  }
+
+  using Comparator::CompareWithoutTimestamp;
+  int CompareWithoutTimestamp(const Slice& a, bool a_has_ts, const Slice& b,
+                              bool b_has_ts) const override {
+    return user_comparator_->CompareWithoutTimestamp(a, a_has_ts, b, b_has_ts);
```
Comment (siying) [inline util/user_comparator_wrapper.h]: Should we bump up counter for user_key_comparison_count as we do in line 33?

File: include/rocksdb/iterator.h [hunk 10]
```diff
@@ -108,6 +108,11 @@ class Iterator : public Cleanable {
...
+    assert(false);
+    return Slice();
+  }
```
Comment (siying) [inline include/rocksdb/iterator.h]: Reading the interface, I'm not very clear when calling Seek(), a user should use key with timestamp or not.

It may help to write some example code in code comments, as what is suggested in https://google.github.io/styleguide/cppguide.html#Function_Comments

Comment (riversand963) [reply include/rocksdb/iterator.h]: Will add comment. Seek(target), target does not contain timestamp. Timestamp is specified in read_options.

File: include/rocksdb/iterator.h (hunk 1 above)
Comment (riversand963) [reply include/rocksdb/iterator.h]: Sure, let me make the change and include in this PR.

File: util/user_comparator_wrapper.h (hunk 9 above)
Comment (riversand963) [reply util/user_comparator_wrapper.h]: Yes we should.

File: db/db_iter.cc (hunk 8 above)
Comment (riversand963) [reply db/db_iter.cc]: This is a good point. Since `sequence_` can also come from user, it is not a good practice to assert on its value. I will remove the assertion. It will be tricky if user specifies both timestamp and sequence number.

Comment (BradBarnich) [reply db/db_iter.cc]: I had to change this to:
```c
if (sequence <= sequence_) {
  return true;
}
return false;
```

otherwise this line failed the assertion: https://github.com/BradBarnich/rocksdb/commit/00d7c3107d81920d38f879b9a5e9848bce7159df#diff-4b40bd5d318908a16d60d610ff1974eaR3311

(which is a copy of IterWithSnapshot with user timestamps, but it relies on reverse iteration support.)

There might be a bug in my reverse iteration support though

Comment (riversand963) [reply db/db_iter.cc]: Do you mean you have to remove the entire comparison of timestamp from the `IsVisible` method? I believe the `IsVisible` should consider timestamp too. I am not familiar with reverse iterator yet, but I think a fix may be needed elsewhere.

**Comment by riversand963** [comment]
One of the appveyor test failure seemed to be caused by timeout. The newly-added timestamp test takes too much time on MSVC in Debug mode.

File: db/db_iter.cc (hunk 8 above)
Comment (BradBarnich) [reply db/db_iter.cc]: No, just replaced the assertion line with this if statement.

File: include/rocksdb/iterator.h [hunk 11]
```diff
@@ -53,11 +54,13 @@ class Iterator : public Cleanable {
   // All Seek*() methods clear any error status() that the iterator had prior to
   // the call; after the seek, status() indicates only the error (if any) that
   // happened during the seek, not any past errors.
+  // Target does not contain timestamp.
```
Comment (siying) [inline include/rocksdb/iterator.h]: I don't see a similar comment to InternalIterator. I think we should comment the same there.

Comment (riversand963) [reply include/rocksdb/iterator.h]: Sure, will add a comment. Note that internal iterator seek target does contain user timestamp.

**Comment by riversand963** [comment]
> I think line 401 to 418 need to be modified. The reseeking logic probably will not work and may cause infinite loop.

Thanks @siying for the review. I have updated the PR.

File: db/db_iter.cc [hunk 12]
```diff
@@ -388,8 +400,17 @@ bool DBIter::FindNextUserEntryInternal(bool skipping_saved_key,
...
         // user-key with decreasing sequence numbers. Fast forward to
         // sequence number 0 and type deletion (the smallest type).
-        AppendInternalKey(&last_key, ParsedInternalKey(saved_key_.GetUserKey(),
-                                                       0, kTypeDeletion));
+        if (timestamp_size_ == 0) {
+          AppendInternalKey(
+              &last_key,
+              ParsedInternalKey(saved_key_.GetUserKey(), 0, kTypeDeletion));
+        } else {
+          std::string min_ts(timestamp_size_, static_cast<char>(0));
```
Comment (siying) [inline db/db_iter.cc]: We assume that all 0 string is the smallest timestamp. Is it fair to make the assumption? Or should we return a min timestamp in comparator?

File: table/internal_iterator.h [hunk 13]
```diff
@@ -52,6 +52,7 @@ class InternalIteratorBase : public Cleanable {
   // All Seek*() methods clear any error status() that the iterator had prior to
   // the call; after the seek, status() indicates only the error (if any) that
   // happened during the seek, not any past errors.
+  // 'target' contains user timestamp if timestamp is enabled.
```
Comment (siying) [inline table/internal_iterator.h]: Make sure Get() follows this rule when Seek() is called in BlockBasedTable::Get().

File: db/db_iter.cc (hunk 12 above)
Comment (riversand963) [reply db/db_iter.cc]: I think in usual cases it is ok, but I can add a new virtual function to `Comparator`, e.g. `Comparator::GetMinTimestamp() const`. I have only one concern. I imagine we are going to add a `clock` object as db options in the future. Should `GetMinTimestamp()` be part of `clock` or `comparator`.

File: table/internal_iterator.h (hunk 13 above)
Comment (riversand963) [reply table/internal_iterator.h]: Current and original assumption is that the `const Slice& key` argument given to `BlockBasedTable::Get()` is internal key, which contains timestamp if timestamp is enabled. Do you mean we should assert the `key.size()` must be greater than or equal to `8 + timestamp_size_`?

File: db/db_iter.cc (hunk 12 above)
Comment (siying) [reply db/db_iter.cc]: Then maybe add it to Clock eventually. We can leave it as 0 now and have a comment here.

File: table/internal_iterator.h (hunk 13 above)
Comment (siying) [reply table/internal_iterator.h]: I don't mean assertion. I just double check that whether it is consistent to how we already use internal iterator.

File: include/rocksdb/comparator.h [hunk 14]
```diff
@@ -97,15 +100,27 @@ class Comparator {
...
+  int CompareWithoutTimestamp(const Slice& a, const Slice& b) const {
+    return CompareWithoutTimestamp(a, /*a_has_ts=*/true, b, /*b_has_ts=*/true);
   }
 
+  // For two events e1 and e2 whose timestamps are t1 and t2 respectively,
+  // Returns value:
+  // < 0  iff t1 < t2
+  // == 0 iff t1 == t2
+  // > 0  iff t1 > t2
+  // Note that an all-zero byte array will be the smallest (oldest) timestamp
```
Comment (riversand963) [inline include/rocksdb/comparator.h]: Added comment saying that all-zero byte array is the smallest (oldest) timestamp.

**Comment by riversand963** [comment]
Thanks @siying for the review!

File: db/db_iter.cc (hunk 12 above)
Comment (riversand963) [reply db/db_iter.cc]: SGTM.
